### PR TITLE
support unregister with register and useController

### LINKED ISSUE #4355: Feature Request - auto unregister inputs on unmount
**Is your feature request related to a problem? Please describe.**
Yes. In V7 and going forward, inputs are not unregistered on unmount. So in other words, we must manually unregister an input if we want its value removed from form values. While this is a great design choice and in most cases makes things much easier, there are cases where we do want form values removed on unmount. 

For example, take a check box that adds additional fields. If the user unchecks the box, then in most cases we can assume that the fields are no longer relevant and we don't want to keep the values around. While it's totally doable to unregister the fields via `watch` and `useEffect`, it adds extra code and logic. It gets especially bad when we have a chain of related fields. For example show C if B and show B if A. We then need to add a bunch of checks to make sure we are not unregistering the field when we shouldn't be. 

**Describe the solution you'd like**
I propose adding a prop to Controller and an option to register that will unregister the input when on unmount. So essentially the V6 behavior. This would give us complete control over how to manage our inputs regarding mounting and unmounting. 

**Describe alternatives you've considered**
Using `useEffect`, `watch`, and `unregister`.

### PR DESCRIPTION
**Pull request by bluebill1049**
close #4355

### CODE REVIEW

File: src/useForm.ts [hunk 1]
```diff
@@ -947,7 +948,8 @@ export function useForm<
             ref: (ref: HTMLInputElement | null) =>
               ref
                 ? registerFieldRef(name, ref, options)
-                : options && options.shouldUnregister && unregister(name),
+                : shouldUnregister ||
+                  (options && options.shouldUnregister && unregister(name)),
```
Comment (iamacook) [inline src/useForm.ts:952]: If `shouldUnregister` is `true`, it won't call `unregister(name)` here.

```js
(shouldUnregister || options && options.shouldUnregister) && unregister(name),
```

Comment (bluebill1049) [reply src/useForm.ts:952]: yep, that's intentional. It's either.

Comment (iamacook) [reply src/useForm.ts:952]: Oh, right - sorry!

Comment (bluebill1049) [reply src/useForm.ts:952]: all good, thanks for the review. If you don't mind, i will ping you for the final review.

Comment (iamacook) [reply src/useForm.ts:952]: That would be great - thank you!

Comment (bluebill1049) [reply src/useForm.ts]: Hey, iamacook. Ready to be reviewed. 👍

Comment (iamacook) [reply src/useForm.ts]: All looks good to me. I'm excited for this feature!

**Pull request merged**